src/mo_frontend/typing.ml

```diff
@@ -2183,31 +2188,149 @@ and infer_call env exp1 inst exp2 at t_expect_opt =
       if not env.pre then check_exp_strong env t_arg' exp2;
       ts, t_arg', t_ret'
     | _::_, None -> (* implicit, infer *)
-      let t2 = infer_exp env exp2 in
+      (*
+        Partial Argument Inference:
+        We need to infer the type of the argument and find the best instantiation for the call expression.
+        However, some expressions cannot be inferred, e.g. unannotated lambdas like `func x = x + 1`.
+        Idea:
+        - Decompose the argument into sub-expressions and defer inference for those that would fail.
+        - Find a partial instantiation first using the inferred sub-expressions.
+        - Substitute and proceed with the remaining sub-expressions to get the full instantiation.
+       *)
+      let infer_subargs_for_bimatch_or_defer env exp target_type =
+        let rec cannot_infer_pat pat =
+          match pat.it with
+          | WildP
+          | VarP _
+          | AltP _ -> true (* cases that cannot be inferred, would report errors *)
+          | LitP _
+          | AnnotP _ (* annotated patterns can be inferred *)
+          | SignP _ -> false
+          | TupP pats -> List.exists cannot_infer_pat pats
+          | ParP p
+          | OptP p
+          | TagP (_, p) -> cannot_infer_pat p
+          | ObjP pfs -> List.exists (fun (pf : pat_field) -> cannot_infer_pat pf.it.pat) pfs
+        in
+        let try_infer_exp env exp =
+          match exp.it with
+          | FuncE (_, _, _, pat, _, _, _) when cannot_infer_pat pat -> None
+          (* Future work: more cases *)
+          | _ -> Some (infer_exp env exp)
+        in
+        let infer_or_defer env (subs, deferred, to_fix) exp target_type =
+          match try_infer_exp env exp with
+          | Some t -> t, ((t, target_type) :: subs, deferred, to_fix) (* subtype problem for bi_match *)
+          | None -> target_type, (subs, (exp, target_type) :: deferred, to_fix) (* deferred *)
+        in
+        let rec decompose env exp target_type acc =
+          match exp.it, T.promote target_type with
```

Comment:
promoting the target_type will replace a type parameter with its bound, are you sure this is what you want here?

Reply:
I just need to unwrap the `Named`, I'll add a separate method that does only that

Reply:
Good catch! Thanks. Done